### DISCUSSION TITLE
HTTP Response validation Error

### DISCUSSION THREAD

**Discussion by fengqi** [dani-garcia/vaultwarden#5287] [Q&A] [answered]
Server Installed Ok
1.32.6
Web Installed Ok
2024.6.2c

I am confused whether to give x-frame-options header or not

```
HTTP Response validation Error

API calls:
Header: 'x-frame-options' is missing!
2FA Connector calls:
Header: 'x-frame-options' is present while it should not

**Comment by BlackDex** [comment, marked as answer]
That depends on the call.
For API Calls it should, for the 2FA `*-connector.html` files it should not.

You actually should not set that header your self.
Remove that from your reverse proxy config and let Vaultwarden handle these headers.

**Reply by fengqi** [reply]
This issue appears to be caused by Nginx. But i reviewing the Nginx configuration, hav't settings related to X-Frame-Options.

**Comment by fengqi** [comment]
nginx configure

```
upstream vaultwarden-default {
    zone vaultwarden-default 64k;
    server bitwarden.xxx.io;
    keepalive 2;
}

map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      "";
}

server {
    listen 443 ssl http2;
    server_name bitwarden.xxxx.com;

    ssl_certificate ssl/xxx.full.crt;
    ssl_certificate_key ssl/xxx.key;
    #more_clear_headers "X-Frame-Options";

    client_max_body_size 128M;

    location / {
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://vaultwarden-default;
    }

}

**Comment by fengqi** [comment]
### Your environment (Generated via diagnostics page)

* Vaultwarden version: v1.32.6
* Web-vault version: v2024.6.2c
* OS/Arch: linux/x86_64
* Running within a container: false (Base: Not applicable)
* Database type: SQLite
* Database version: 3.47.1
* Environment settings overridden!: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Websocket Check: true
* HTTP Response Checks: false

### Config & Details (Generated via diagnostics page)

<details><summary>Show Config & Details</summary>

**Failed HTTP Checks:**

```yaml
2FA Connector calls:
Header: 'x-frame-options' is present while it should not
```

**Config:**

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": false,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_max_note_size": 10000,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*******************",
  "domain_origin": "*****://*******************",
  "domain_path": "",
  "domain_set": true,
  "duo_context_purge_schedule": "30 * * * * *",
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "duo_use_iframe": false,
  "email_2fa_auto_fallback": false,
  "email_2fa_enforce_on_verified_invite": false,
  "email_attempts_limit": 3,
  "email_change_allowed": true,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "enable_websocket": true,
  "enforce_single_org_with_reset_pw_policy": false,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "experimental_client_feature_flags": "fido2-vault-credentials",
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "http_request_block_non_global_ips": true,
  "http_request_block_regex": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 3,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "increase_note_size_limit": false,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/var/log/bitwarden.log",
  "log_level": "warn",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": true,
  "push_identity_uri": "https://identity.bitwarden.com",
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": false,
  "sends_folder": "data/sends",
  "show_password_hint": true,
  "signups_allowed": false,
  "signups_domains_whitelist": "*********,*********",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": false,
  "smtp_from": "******************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "**************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": true,
  "smtp_timeout": 15,
  "smtp_username": "******************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "user_send_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```

</details>

**Reply by BlackDex** [reply]
Well, somewhere in-between it has to add this header.

Check your nginx config again with something like this.
```bash
# Adjust the path to the main nginx config directory so it will search all nginx config files
# Default is /etc/nginx
grep -Iir 'x-frame-options' /path/to/nginx/
```

**Reply by fengqi** [reply]
Thank you for help.

I know why, because before I modified the nginx configuration, I opened Chrome's devtools, which caused the page to be cached.

**Comment by darkman242** [comment]
I have the same problem with version 1.32.6/.7. The following messages appear on the diagnostic screen

2FA Connector calls:
Header: 'x-frame-options' is present while it should not
Header: 'content-security-policy' is present while it should not

These messages do not appear with version 1.32.5. 

I use SWAG as a reverse proxy. Are the Locations API and NotificationHub still required in the current NGINX config file?

https://github.com/linuxserver/reverse-proxy-confs/blob/master/vaultwarden.subdomain.conf.sample

**Reply by BlackDex** [reply]
Those checks are new, so that isn't strange. 
Having those headers set could cause issue.

Vaultwarden already sets all the headers which are needed when they are needed. So do not set them via a reverse proxy.

**Comment by darkman242** [comment]
Sorry to ask, but what exactly do I have to do?
If I set the following settings in the ProxyConf under location and location/admin, the following error messages are displayed?
- proxy_hide_header X-Frame-Options;
- proxy_hide_header Content-Security-Policy;

<img width="1253" alt="image" src="https://github.com/user-attachments/assets/67165dfd-49f4-4434-ae5f-3360b527a65f" />

And are the Locations API and NotificationHub still required in the current NGINX config file?

**Reply by BlackDex** [reply]
It looks like you set those headers your self in your nginx config. This is not needed and probably not wise either.
Vaultwarden already returns all the security headers which are needed when they are needed.

Check the wiki https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples on how to configure your environment.

**Reply by darkman242** [reply]
We have activated the optional headers in SWAG in ssl.conf. In the Proxy.conf for Vaultwarden, we have set the headers for the locations as follows.

add_header Content-Security-Policy "";
add_header X-Frame-Options "";

@BlackDex It works as desired. Thank you

**Comment by artasoftware** [comment]
Having the same:

2FA Connector calls:
Header: 'x-frame-options' is present while it should not

I haven't set this header anywhere. If I add "proxy_hide_header X-Frame-Options;" in the nginx config file, then I see:

API calls:
Header: 'x-frame-options' is missing!
2FA Connector calls:
Header: 'x-frame-options' is present while it should not

My config is: nginx -> docker-compose VW. Nothing more.

**Reply by BlackDex** [reply]
It must be set somewhere, it's not being set out of the blue.

And adding the hide header proved that it is set somewhere else for some specific types somewhere.

Could security plugins of nginx like ModSecurity or WAF.

**Reply by artasoftware** [reply]
> It must be set somewhere, it's not being set out of the blue.

Yeah but there is no place to be set. I do see "x-frame-options: SAMEORIGIN" in the developer tools' headers but I can't find it anywhere. 

> And adding the hide header proved that it is set somewhere else for some specific types somewhere.
Why? When I added "proxy_hide_header X-Frame-Options;", API calls got broken but 2FA Connector was the same as before setting. 

Maybe it is possible to set "proxy_hide_header X-Frame-Options;" for 2FA connector calls location or endpoint or whatever?

**Reply by BlackDex** [reply]
Your last option will not help since Vaultwarden does not send those headers for those endpoints, and thus proxy hide shouldn't help there unless you have another proxy or tunnel inbetween

**Reply by artasoftware** [reply]
> Your last option will not help since Vaultwarden does not send those headers for those endpoints, and thus proxy hide shouldn't help there unless you have another proxy or tunnel inbetween

I can't understand then why it appeared. I've update from 1.32.2 to 1.32.7 today and it appeared. Configuration wasn't changed.

**Reply by BlackDex** [reply]
It is very simple, it is a new feature since v1.32.6 and more specifically #5272
Before that version this check was not there, and it could have caused strange issues in regards of using some 2FA options using other clients then the web-vault.

That is why those checks are added, including if reverse proxy return something else then the errors Vaultwarden is sending, which also breaks clients.

**Comment by derritter88** [comment]
@BlackDex thanks for this thread - saved my butt today!
Kudos to you!! 👍

**Comment by skyreacher** [comment]
I had the same issue, using traefik (3.4.1) as reverse proxy. My problem was that I had some "default headers" defined for security reasons in a middleware, which I used with all services running in Docker containers.

So what helped was not to use these headers with vaultwarden, as often suggested in this discussion (see [https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples ](https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples) for a setup with traefik). Afterwards Vaultwarden's diagnostic page still showed an error, but it looks like that was just a browser caching issue - force reloading the page helped, and now it shows all checks green.

**Comment by sebtech33** [comment]
Using Traefik v3.4.1 I still get errors, even when removing every middleware and every header modifications.
Here is the following error's i get when looking at the diagnostics page. Yes i have force reloaded the page and tried to use another web browser and it still shows this:

HTTP Response validation Error
2FA Connector calls:
Header: 'x-xss-protection' does not contain '0'
Header: 'x-robots-tag' does not contain 'noindex'
Header: 'x-robots-tag' does not contain 'nofollow'
Header: 'x-frame-options' is present while it should not

**Reply by sebtech33** [reply]
NVM I seem to have fixed it.

**Reply by artasoftware** [reply]
How did you fix it?

**Reply by sebtech33** [reply]
@artasoftware I fixed it by making vaultwarden handle all headers itself. Removing all middlewares (also security headers) and exposing it. Quick tip also: Inspect the web page - under networking tab - disable caching (force reload did not remove the errors only disabling the cache on the website. Now I see no errors. Sorry for a messy reply, writing from memory.
